Disable implicit actions if update query to include clauses

qnkhuat opened the pull request:
Resolves #28949

Implicit actions are only allowed on a simple model (aka model with only a source table).
This PR will make
- update model query to include new clauses (sort, aggregation, etc ) will need remove any implicit actions that are linked to it
- update the API to make sure we don't allow creating implicit actions on a complex model. (we already had this check on FE so it's not possible to hit this from UI, but add it just to make sure)

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/metabase/metabase/29021)
<!-- Reviewable:end -->

kulyk committed:
Fix dashboard filter values coercion for actions (#28963)

* Fix dashboard filter values coercsion for actions

* update custom action e2e test to cover parameters

---------

Co-authored-by: Ryan Laurie <iethree@gmail.com>

qnkhuat committed:
the logic

metamben committed:
MLv2 filter schema (#28912)

* Make nil equality comparable
* Add Malli schema for filter expressions

qnkhuat committed:
add tests

qnkhuat committed:
remove stray calls

qnkhuat committed:
fix serdes tests

qnkhuat committed:
add test for an edge case

kulyk committed:
Hide new question link without data permissions (#29022)

kulyk committed:
Fix actions behavior on the archive page (#28994)

* Properly connect actions to search entity

* Add e2e test

qnkhuat committed:
db changes tracking utilities (#28256)

ranquild committed:
Handle permissions with actions correctly (#28990)

qnkhuat committed:
fix serdes test and add a check for create API

qnkhuat commented on src/metabase/models/card.clj:
```diff
@@ -344,8 +361,13 @@
+        (disable-implicit-action-for-model! id))
       ;; Archive associated actions
-      (when (and (not (:dataset changes))
+      (when (and (false? (:dataset changes))
```
this was a bug before and tested it here : https://github.com/metabase/metabase/blob/fe252f7c79a8584b18a10a195d9007acd8983a6b/test/metabase/models/action_test.clj#L137

qnkhuat commented on enterprise/backend/test/metabase_enterprise/serialization/v2/e2e/yaml_test.clj:
```diff
@@ -158,6 +158,16 @@
                                                {:table_id      [:t    100]
                                                 :collection_id [:coll 100]
                                                 :creator_id    [:u    10]}))
+               :simple-model            (mapv #(update-in % [1 :refs] table->db)
```
with the new `pre-update` in card, all implicit actions will be deleted during the generate part( I'm not sure why it calls update here but it is what it is).

The reason implicit actions got deleted is that the spec for `:card` in this test is a complex query with aggregation and breakout.

So I need to create a new spec that use a simple model so that our actions are safe from updates.

calherries replied:
Can you add a comment for this in the code? It would be almost impossible to know this without diving into this PR otherwise.

qnkhuat replied:
yep, added here: https://github.com/metabase/metabase/pull/29021/commits/1d8c6582972870b47b21603782d8618639ef371d

nllho committed:
docs - update public sharing (#29001)

qnkhuat commented on src/metabase/models/card.clj:
```diff
@@ -317,12 +317,29 @@
+  meaning there are no clauses such as(filter, limit, breakout...)"
+  [{dataset-query :dataset_query :as _card}]
+  (and (= :query (:type dataset-query))
+       (every? #(nil? (get-in dataset-query [:query %])) [:expressions :filter :limit :breakout :aggregation :joins])))
```
it matches the check on FE: https://github.com/metabase/metabase/blob/89f8d515fe544cf85ed810b0b4cdc112222c71b5/frontend/src/metabase-lib/queries/StructuredQuery.ts#L508

calherries replied:
Again, this kind of comment should live in the code. Whenever there's a co-dependency between BE and FE there should be a comment in both spots, so if the FE updates they know the BE needs to update as well, and vice versa.

calherries replied:
This also doesn't match the FE perfectly. You need to add `:order-by` and `:fields`
```suggestion
       (every? #(nil? (get-in dataset-query [:query %])) [:expressions :filter :limit :breakout :aggregation :joins :order-by :fields])))
```

qnkhuat replied:
nice catch, Added the comment and include the missing clauses here: https://github.com/metabase/metabase/pull/29021/commits/1d8c6582972870b47b21603782d8618639ef371d

calherries commented on src/metabase/models/card.clj:
```diff
@@ -344,8 +361,13 @@
       ;; make sure this Card doesn't have circular source query references if we're updating the query
       (when (:dataset_query changes)
         (check-for-circular-source-query-references changes))
+      ;; updating a model dataset query to includes any clauses will disable implicit acitons(if it exists)
```
```suggestion
      ;; updating a model dataset query to not support implicit actions will disable implicit actions if they exist
```

calherries commented on src/metabase/models/card.clj:
```diff
@@ -344,8 +361,13 @@
-      (when (and (not (:dataset changes))
+      (when (and (false? (:dataset changes))
                  (:dataset old-card-info))
         (t2/update! 'Action {:model_id id} {:archived true}))
```
Not related to your PR but I just realised this is a bug too. Implicit actions should be deleted when a card is changed from a model to a card. Currently they are archived. Only query actions should be archived.

qnkhuat replied:
nice catch. you'll fix it, right?

calherries replied:
yes

calherries commented on test/metabase/test/generate.clj:
```diff
@@ -238,6 +238,14 @@
                                               :database_id   [:database :id]
                                               :table_id      [:table :id]
                                               :collection_id [:collection :id]}}
+   ;; like card but is a model and the query is very simple
```
Would be good to document why we need this here as well

qnkhuat replied:
added a line under this : https://github.com/metabase/metabase/pull/29021/commits/1d8c6582972870b47b21603782d8618639ef371d

calherries commented on test/metabase/models/card_test.clj:
```diff
@@ -107,6 +109,67 @@
 
+(deftest disable-implicit-actions-if-needed-test
+  (mt/with-actions-enabled
+    (testing "when updating a model to include any clauses will disable implicit actions if exists\n"
```
```suggestion
    (testing "when updating a model to include any clauses will disable implicit actions if they exist\n"
```

iethree committed:
Normalize Link Card icon sizes and spacing (#28926)

* normalize icon sizes and spacing

* clean up some types

iethree committed:
Improved datetime handling in Action Forms 🥴  (#29002)

* improved datetime handling

* 💪 more tests 💪

markbastian committed:
Minor.Point Dependency Updates for Cycle 47 (#28969)

All minor version dependencies have been updated to the latest. 3 libraries caused breakage and are documented here: https://github.com/metabase/metabase/issues/28967.

camsaul committed:
Mention changes to building drivers in Metabase 0.46.0 (#29045)

ranquild committed:
Fix copy-pasting sql into the action editor (#29034)

EmmadUsmani committed:
replace modal for removing dashcards with undo toast (#28970)

snoe committed:
Re-apply Remove usages of dataset_query.type (#28957)

* Remove usages of dataset_query.type (#28683)

* Remove usages of dataset_query.type

* Remove more usages of dataset_query

* Introduce usesSegment and usesMetric methods on Question

* Fix type check

* Fix type

* Get query for UtilQuery

* Fix conflicting added imports

dpsutton committed:
Fix flaky json test (#29048)

Occasionally in CI we get

```
“Syntax error compiling at (release/version_info_test.clj:46:26).\nNo such var: json/read-json\n”,
```

from `release.version-info-test`.
This namespace requires `[clojure.data.json :as json]` and uses the
function `json/read-json` which is deprecated since version 2.x.

The way that deprecated function is defined in version 2.0.2 is

```clojure
(ns clojure.data.json)
...
(load "json_compat_0_1")
```

and in `clojure/data/json_compat_0_1.clj`:

```clojure
(in-ns 'clojure.data.json)

(defn read-json
  "DEPRECATED; replaced by read-str.
  ...)
```

And I guess the way that hawk requires some namespaces we can get into
this half loaded, half not finished state and the var is not (yet)
defined.

So we can just put aside this nonsense and use a non-deprecated var that
isn't defined with a `load` statement but in the regular namespace we
require.

npfitz committed:
Adding classname prop to ObjectDetail (#29032)

* Adding classname prop to ObjectDetail

* making className optional

alxnddr committed:
add explicit method param for saved and unsaved questions (#29024)

camsaul committed:
Misc MLv2 improvements from #28921 (#29010)

* MLv2 metadata overhaul, second try

* Stricter linting for MLv2

* Address PR feedback

* Revert fancy fn schema for now

* Remove unused import

* Test fixes :wrench:

* Oops fix bad :require form

* Misc MLv2 improvements from #28921

* Fix indentation

* expression.non-integer-real

* Remove debugging stuff

* More cleanup

* Test fixes: ::ref/field shouldn't fail if checked against a non-sequential object

qnkhuat committed:
more docs and add order-by, sort to the list

qnkhuat committed:
fix for mysql and mariadb

ranquild committed:
Do not show parameters for non-variable template tags in the action editor (#29028)

kulyk committed:
Fix docs link in action dashcard settings (#29046)

qnkhuat committed:
resolve comments and un comment skipped tests

calherries committed:
Remove unnecessary `retrieve-dashboard-card` call on every update to a dashcard (#29037)

* Remove retrieve-dashboard-card call

* Have `update-dashboard-card!` return nil, not true

calherries committed:
Delete implicit actions when models are changed to saved questions (#29078)

dpsutton committed:
Remove personal collections from collection detail page (#28879)

* Remove personal collections from collection detail page

https://github.com/metabase/metabase/pull/28023 Added "sub collections"
to the collection detail page. Makes sense. The problem is that personal
collections started appearing in this view. We have traditionally
treated personal collections as siblings to the "Our Analytics" root of
collections. So this was an impedence mismatch.

This PR just removes all personal collections from the items returned
from `"api/collection/root/items"`. The personal collection is still
visible in the sidebar (returned from
`api/collection/tree?tree=true&exclude-other-user-collections=true&exclude-archived=true`
. Furthermore, all personal collections are still visible from the
"three dots" menu on the sidebar.

* poke ci

* remove last personal collections from tests

* e2e tests

don't include a personal collection in the details page.

* off by one from removing personal collections

* Can't select personal collections if they aren't there

qnkhuat committed:
Adds `metabase_field.database_auto_incremented` (#28942)

npfitz committed:
pulling out column title logic and sharing with settings (#28937)

* pulling out column title logic and sharing with settings

* Pivot Table Custom Column Names

* removing cyclic dependency

* adjusting tests

kulyk committed:
Better handle numeric action parameters with "select" and "radio" types (#29000)

* Don't ignore manually entered parameters on clean

* Validate options for numeric fields with dropdown input

* Clear `valueOptions` when switching field and input types

* Remove sample options for inputs with options

* Rename `getValidationError`

* Bring sample options back

* Minor rename for clarity

* Fix `InputOptionsMap` type

* Keep `inputType` on field type change if possible

* Extract `FieldValueOptions` type

* Use less aggressive value options clean logic

* Simplify tests

* Test value options reset for option-less field types

* Fix `OptionEditor` value gets out of sync

tsmacdonald commented on frontend/src/metabase-lib/queries/StructuredQuery.ts:
```diff
@@ -506,6 +506,7 @@ class StructuredQueryInner extends AtomicQuery {
   }
 
   hasAnyClauses() {
+    // this list should be kept in sync with BE in `metabase.models.card/model-supports-implicit-actions?`
```
Candidate for (eventual) CLJS-ification?

qnkhuat replied:
yep

tsmacdonald commented on src/metabase/models/card.clj:
```diff
@@ -317,12 +317,34 @@
 
+(defn model-supports-implicit-actions?
+  "A model with implicit action supported iff they are a raw table,
+  meaning there are no clauses such as(filter, limit, breakout...).
```
```suggestion
  meaning there are no clauses such as filter, limit, breakout...
```

tsmacdonald reviewed (approved):
:clap:

john-metabase committed:
Serdes v2 namespace reorg (#29066)

* Reorganizes serdes v2 namespaces and some functions/multimethods to simplify implementation
* Updates serdes v2 to store/load :serdes/meta in YAML to remove ingestion path matching

markbastian committed:
Update snowplow-java-tracker dependency to 1.0.0 (#29042)

Upgraded com.snowplowanalytics/snowplow-java-tracker from 0.12.0 to 1.0.0. This did require refactoring the `metabase.analytics.snowplow` ns to use the new fluent API as well as suppressing eastwood linting of the ns due to a classloader issue as described [here](https://github.com/jonase/eastwood/issues/444).

dragonsahead committed:
Add creation of user on the fly JWT (#28375)

camsaul committed:
MLv2 clause and type determination overhaul (#29070)

* MLv2 `define-mbql-clause` and `type-of` calculation

* Continued [ci skip]

* MLv2 overhaul :flex:

camsaul committed:
MLv2 metadata calculation (#28921)

* MLv2 metadata overhaul, second try

* Stricter linting for MLv2

* Address PR feedback

* Revert fancy fn schema for now

* Remove unused import

* Test fixes :wrench:

* Oops fix bad :require form

* MLv2 mega metadata overhaul 2 of infinity

* Test fixes

* Cljc humanization

* Misc MLv2 improvements from #28921

* Fix indentation

* expression.non-integer-real

* Remove debugging stuff

* More cleanup

* Add method for `:advanced`

* non-integer-real

* Test fixes :wrench:

* Test fixes: ::ref/field shouldn't fail if checked against a non-sequential object

* MLv2 `define-mbql-clause` and `type-of` calculation

* Continued [ci skip]

* MLv2 overhaul :flex:

* Address PR feedback

* Merge type-of overhaul

* Address PR feedback

markbastian committed:
Update org.slf4j/slf4j-api from 1.7.36 to 2.0.6 (#29091)

Updated both the slf4j api and impl jars.

qnkhuat committed:
Apply suggestions from code review

Co-authored-by: Tim Macdonald <tim@metabase.com>

qnkhuat committed:
Merge branch 'master' into ngoc-disable-implicit-actions-if-update-query

qnkhuat commented:
@calherries could you help re-review this?

qnkhuat committed:
miss a dot, fail CI :(

qnkhuat committed:
Merge branch 'master' into ngoc-disable-implicit-actions-if-update-query

qnkhuat committed:
GH resolve conflicts editor is 😡

calherries reviewed (approved):
Nice
